<a href="https://colab.research.google.com/github/pondbaahh/DWDM21/blob/main/MiniExam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### ข้อสอบ

จาก data ```flights.csv``` และ ```airports.csv``` ใน https://drive.google.com/drive/folders/1EFo1_uQgWkRiXjezL0T4YwteYrsqU5MV?usp=sharing
0. หารัฐที่มีจำนวนเครื่องบินขาเข้ามากที่สุดและน้อยที่สุด
1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด
2. วาด Histogram แสดงจำนวนเครื่องบินขาออกของแต่ละวันในสัปดาห์
3. สายการบินไหนมีเที่ยวบินมากที่สุดและน้อยที่สุด
4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด
5. ช่วงเช้าหรือช่วงบ่ายมีเครื่องบินขาเข้ามากกว่ากัน

ให้นักศึกษาทำข้อสอบ 2 ข้อจากที่แสดงข้างบนโดยกรอกรหัสนักศึกษาลงใน function ```gen_exams()``` แล้ว run เพื่อให้ได้ข้อสอบที่นักศึกษาต้องทำ 2 ข้อ (ข้อละ 3 คะแนน)
และให้นักศึกษาคิดโจทย์เองอีก 1 ข้อ (4 คะแนน)


In [1]:
import pandas as pd

In [2]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os

In [4]:
flights = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/flights.csv')
flights

วันในเดือน  วันในสัปดาห์  ... ความล่าช้าขาออก  ความล่าช้าขาเข้า
0                19             5  ...              -3                 1
1                19             5  ...               0                -8
2                19             5  ...              -4               -15
3                19             5  ...              28                24
4                19             5  ...              -6               -11
...             ...           ...  ...             ...               ...
2702213          19             6  ...               1                -3
2702214          19             6  ...              -2                -3
2702215          19             6  ...              18                 9
2702216          19             6  ...              10                 7
2702217          19             6  ...               0               -13

[2702218 rows x 7 columns]

In [5]:
airports = pd.read_csv('/content/drive/MyDrive/DWDM2021_DATA/Mini Exam/airports.csv')
airports

รหัสสนามบิน         เมือง รัฐ                          ชื่อสนามบิน
0          10165   Adak Island  AK                                 Adak
1          10299     Anchorage  AK  Ted Stevens Anchorage International
2          10304         Aniak  AK                        Aniak Airport
3          10754        Barrow  AK      Wiley Post/Will Rogers Memorial
4          10551        Bethel  AK                       Bethel Airport
..           ...           ...  ..                                  ...
360        11233      Cheyenne  WY  Cheyenne Regional/Jerry Olson Field
361        11097          Cody  WY                 Yellowstone Regional
362        11865      Gillette  WY             Gillette Campbell County
363        12441       Jackson  WY                         Jackson Hole
364        14543  Rock Springs  WY       Rock Springs Sweetwater County

[365 rows x 4 columns]

In [6]:
def gen_exams(ID):
    import math
    a = int(ID[-1])%6
    b = int(ID[-1])%3
    print(f'รหัสนักศึกษา {ID} ทำข้อ {a} และ {b if a != b else 5-b}')

ตัวอย่างการ run

In [7]:
gen_exams('62123456-9')

รหัสนักศึกษา 62123456-9 ทำข้อ 3 และ 0


In [8]:
gen_exams('623020535-7')

รหัสนักศึกษา 623020535-7 ทำข้อ 1 และ 4


#หา Missing value

In [9]:
flights.isnull().any()

วันในเดือน            False
วันในสัปดาห์          False
สายการบิน             False
รหัสสนามบินต้นทาง     False
รหัสสนามบินปลายทาง    False
ความล่าช้าขาออก       False
ความล่าช้าขาเข้า      False
dtype: bool

In [10]:
airports.isnull().any()

รหัสสนามบิน    False
เมือง          False
รัฐ            False
ชื่อสนามบิน    False
dtype: bool

ตารางทั้งสองไม่มีค่า Missing value

# รวมสองตาราง

In [78]:
merged_table = flights.merge(airports,how='left',left_on='รหัสสนามบินต้นทาง',right_on='รหัสสนามบิน')
merged_table

วันในเดือน  ...                                 ชื่อสนามบิน
0                19  ...                  Detroit Metro Wayne County
1                19  ...                Salt Lake City International
2                19  ...                      Portland International
3                19  ...             Lambert-St. Louis International
4                19  ...  Cincinnati/Northern Kentucky International
...             ...  ...                                         ...
2702213          19  ...                       Orlando International
2702214          19  ...    Hartsfield-Jackson Atlanta International
2702215          19  ...           Albuquerque International Sunport
2702216          19  ...    Hartsfield-Jackson Atlanta International
2702217          19  ...    Hartsfield-Jackson Atlanta International

[2702218 rows x 11 columns]

#1. หารัฐที่มีจำนวนเครื่องบินขาออกมากที่สุดและน้อยที่สุด

In [44]:
Max= merged_table.sort_values(by=['รัฐ'],ascending=False)
Max.iloc[:1,[3,9]]

รหัสสนามบินต้นทาง รัฐ
300811              13487  WI

In [43]:
Min= merged_table.sort_values(by=['รัฐ'],ascending=True)
Min.iloc[:1,[3,9]]

รหัสสนามบินต้นทาง รัฐ
507270              14057  AK

รัฐ WI มีจำนวนเครื่องบินขาออกมากที่สุดและรัฐ AK มีจำนวนเครื่องบินขาออกน้อยที่สุด

#4. สายการบินอะไรมีความล่าช้ามากที่สุดและน้อยที่สุด

In [28]:
Max1 =merged_table.sort_values(by=['ความล่าช้าขาออก'],ascending=False)
Max1.iloc[:1,[2,5]]

สายการบิน  ความล่าช้าขาออก
2632902        AA             1863

In [29]:
Min1 =merged_table.sort_values(by=['ความล่าช้าขาออก'],ascending=True)
Min1.iloc[:1,[2,5]]

สายการบิน  ความล่าช้าขาออก
2576772        F9              -63

In [58]:
Max2 =merged_table.sort_values(by=['ความล่าช้าขาเข้า'],ascending=False)
Max2.iloc[:1,[2,6]]

สายการบิน  ความล่าช้าขาเข้า
1230689        HA              1845

In [60]:
Min2 =merged_table.sort_values(by=['ความล่าช้าขาเข้า'],ascending=True)
Min2.iloc[:1,[2,6]]

สายการบิน  ความล่าช้าขาเข้า
666915        YV               -94

สายการบิน AA มีความล่าช้าขาออกมากที่สุด และสายการบิน F9 มีความล่าช้าขาออกน้อยสุด

สายการบิน HA มีความล่าช้าขาเข้ามากที่สุด และสายการบิน YV มีความล่าช้าขาเข้าน้อยสุด

#โจทย์คิดเอง

รัฐไหนที่มีค่าเฉลี่ยของ 'ความล่าช้าขาออกและความล่าช้าขาเข้า' สูงที่สุด 4 อันดับเเรก

In [90]:
pp = merged_table.groupby('รัฐ').mean().sort_values(by=['ความล่าช้าขาออก'],ascending=False)
pp.iloc[:4,[4]]

ความล่าช้าขาออก
รัฐ                 
IL         15.806238
NJ         14.552501
CO         14.463063
MD         13.692181

รัฐที่มีค่าเฉลี่ยของ 'ความล่าช้าขาออก' สูงที่สุด 4 อันดับแรก

1. รัฐIL เท่ากับ 15.806238
2. รัฐNJ เท่ากับ 14.552501
3. รัฐCO เท่ากับ 14.463063
4. รัฐMD เท่ากับ 13.692181

In [93]:
pp1 = merged_table.groupby('รัฐ').mean().sort_values(by=['ความล่าช้าขาเข้า'],ascending=False)
pp1.iloc[:4,[5]]

ความล่าช้าขาเข้า
รัฐ                  
IL          11.654502
CO          11.441419
MD           9.631479
PR           8.353468

รัฐที่มีค่าเฉลี่ยของ 'ความล่าช้าขาเข้า' สูงที่สุด 4 อันดับแรก
1. รัฐIL เท่ากับ 11.654502
2. รัฐCO เท่ากับ 11.441419
3. รัฐMD เท่ากับ 9.631479
4. รัฐPR เท่ากับ 8.353468